ber.go

```diff
@@ -9,6 +9,8 @@ import (
 	"reflect"
 )
 
+const maxPacketLengthBytes = 0x8000000 // 134,217,728 bytes (chosen arbitrarily)
```

Comment:
Seems reasonable

Reply:
Ldap search results come back as a single packet, right? I’ve seen some surprisingly large result sets that would definitely exceed this

Reply:
@liggitt I'm looking for some guidance here. What would be reasonable, 1-2GB? We could also make it a package level variable if a user encounters a use can which needs a higher value.

The issue is `make([]byte, length)` panics if there's not enough memory. The intention here is to protect ourselves from an errant server or network which returns an obviously invalid result. What's "obviously invalid" isn't always so obvious.

References for `make()` and its limits:
- https://golang.org/src/runtime/slice.go#L31
- https://golang.org/src/runtime/malloc.go#L137

Reply:
> @liggitt I'm looking for some guidance here. What would be reasonable, 1-2GB? We could also make it a package level variable if a user encounters a use can which needs a higher value.

yeah... it's tricky. there's basically four modes of operating:
* stream results to disk, support partial reads/lazy-loading/on-demand-parsing. requires fundamentally changing the API, not really an option here.
* allow arbitrarily large allocations on a system with paging enabled, allocs larger than available physical memory are (sometimes) tolerated by the underlying OS, other times oom
* allow arbitrarily large allocations on a system without paging enabled, allocs larger than available physical memory oom
* artificially set a max size in the library. this is essentially impossible to do correctly for all environments (in some embedded systems, 100MB could still fail, in other systems, 2GB of search results is perfectly tolerable)

We should also decide what we're protecting against:
* malicious server? if so, they could just set indeterminate length and stream forever anyway, so this doesn't really help.
* spurious data that flips a bit in the length descriptor? if so, we could simply set a max size on the *initial* buffer we accumulate into, but let it grow as we read the data. this would protect against a case where a malformed 1k packet claimed to have a 1TB of data, since we'd EOF and error before actually needing to allocate a buffer big enough to hold that.

I'd be in favor of either of the following:
* max initial buffer size, allow growing while reading packet
* package-level absolute max size, settable by consumers, defaulting to no limit (as today)

Reply:
I like the idea of a package variable.

Reply:
> "sane" max initial buffer size (1MB? 10MB?), allow growing while reading packet

@liggitt @johnweldon I'd prefer this also, but it's additional work which needs to be done carefully. I'd recommend doing this in a follow-up. I'll be upfront, I can't promise I'll be the one to do it 😄 

Maybe using [io.LimitedReader](https://golang.org/pkg/io/#LimitedReader) would help.

The motivation was to limit panics which could crash the application, however they may come about. If we start getting into how that may happen it usually derails the conversation (unless we're specifically doing threat modeling).

> package-level absolute max size, settable by consumers, defaulting to no limit (as today)

I think this is a sane quick fix. In the test it attempts to allocated 200+ GB, which I would assume is outside even most edge case uses.

For now I changed it to an exported variable with a default of `math.MaxInt32` (~2GB). It works on my 64-bit machine, and Travis shouldn't have any issues because there isn't a test for allocating the max allowed byte slice.

Please let me know if this is acceptable or if you'd like to try something else.

Reply:
I think it's acceptable.  Thank you again for your contributions.  I'll make sure @liggitt doesn't have any reservations, and then we'll merge this in.